src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/Utilities.cs

```diff
@@ -88,37 +88,94 @@ internal class ScriptingRuntimeHelpers
     {
         public static object Int32ToObject(int i)
         {
+            switch (i)
+            {
+                case -1:
+                    return Int32_m;
+                case 0:
+                    return Int32_0;
+                case 1:
+                    return Int32_1;
+                case 2:
+                    return Int32_2;
+                case 3:
+                    return Int32_3;
+            }
+
             return i;
         }
+        
+        internal static readonly object Int32_m = -1;
+        internal static readonly object Int32_0 = 0;
+        internal static readonly object Int32_1 = 1;
+        internal static readonly object Int32_2 = 2;
+        internal static readonly object Int32_3 = 3;
```

Comment:
Yes, this is to avoid boxing. That would seem to be the point of Int32ToObject.
Yes the constants can be private.

First I needed only 0 here, but then realized that there are other common cases like 1 and -1 so I have added them and then couple more for 2, 3. 
I think the bigger the number, the smaller is the statistical chance it would be used. It is a complete guess at how many are enough. 
IL constants have range [-1 : 8] , but that at best just an analogy.